src/gt4py/next/otf/workflow.py

```diff
@@ -61,7 +71,82 @@ def __call__(self, inp: StartT_contra) -> EndT_co:
 
 
 @dataclasses.dataclass(frozen=True)
-class Step(Generic[StartT, EndT]):
+class NamedStepSequence(Generic[StartT, EndT]):
+    """
+    Workflow with linear succession of named steps.
+
+    Examples:
+    ---------
+    >>> import dataclasses
+
+    >>> def parse(x: str) -> int:
+    ...    return int(x)
+
+    >>> def plus_half(x: int) -> float:
+    ...    return x + 0.5
+
+    >>> def stringify(x: float) -> str:
+    ...    return str(x)
+
+    >>> @dataclasses.dataclass(frozen=True)
+    ... class ParseOpPrint(NamedStepSequence[str, str]):
+    ...    parse: Workflow[str, int]
+    ...    op: Workflow[int, float]
+    ...    print: Workflow[float, str]
+
+    >>> pop = ParseOpPrint(
+    ...    parse=parse,
+    ...    op=plus_half,
+    ...    print=stringify
+    ... )
+
+    >>> pop.step_order
+    ['parse', 'op', 'print']
+
+    >>> pop(73)
+    '73.5'
+
+    >>> def plus_tenth(x: int) -> float:
+    ...   return x + 0.1
+
+
+    >>> pop.replace(op=plus_tenth)(73)
+    '73.1'
+    """
+
+    def __call__(self, inp: StartT) -> EndT:
+        """Compose the steps in the order defined in the `.step_order` class attribute."""
+        step_result: Any = inp
+        for step in [getattr(self, s) for s in self.step_order]:
+            step_result = step(step_result)
+        return step_result
+
+    def chain(self, step: Workflow[EndT, NewEndT]) -> CombinedStep[StartT, EndT, NewEndT]:
+        return CombinedStep(first=self, second=step)
```

Comment:
```suggestion
    def chain(self, step: Workflow[EndT, NewEndT]) -> CombinedStep[StartT, EndT, NewEndT]:
        return StepSequence(steps=self.step_order+[step])
```

Sketch as discussed (naming and class hierarchy up to you):
```suggestion
class StepSequence(...):
    workflows: list[Workflow]
    
    def chain(step):
        return Steps(workflows=self.workflows+[step])
```

Reply:
A note about the diff between this sketch and my implementation: By disallowing initialization with arbitrary length lists I managed to keep the same level of mypy usefulness as before. Chaining happens using `dataclasses.replace` and `typing.cast`.

If we decide later that all this static typing is not useful, we can replace all that stuff with a couple of `Any`s later and get rid of this restriction.